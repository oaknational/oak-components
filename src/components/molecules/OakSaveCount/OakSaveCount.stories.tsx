import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakSaveCount } from "./OakSaveCount";

const meta: Meta<typeof OakSaveCount> = {
  component: OakSaveCount,
  tags: ["autodocs"],
  title: "components/molecules/OakSaveCount",
  argTypes: {
    count: {
      control: {
        type: "number",
      },
    },
    href: {
      control: {
        type: "text",
      },
    },
    loading: {
      control: {
        type: "boolean",
      },
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakSaveCount>;

export const Default: Story = {
  render: (args) => <OakSaveCount {...args} />,
  args: {
    count: 1,
    href: "#",
    loading: false,
  },
};
