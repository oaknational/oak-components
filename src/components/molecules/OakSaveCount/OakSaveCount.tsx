import React from "react";
import styled from "styled-components";

import {
  OakBox,
  OakFlex,
  OakIcon,
  OakScreenReader,
  OakSpan,
} from "@/components/atoms";
import { InternalButton } from "@/components/atoms/InternalButton";
import { parseColor } from "@/styles/helpers/parseColor";
import { parseDropShadow } from "@/styles/helpers/parseDropShadow";

export type OakSaveCountProps = {
  count: number;
  href: string;
  loading: boolean;
};

const StyledInternalButton = styled(InternalButton)`
  :hover {
    .oak-save-count {
      background-color: ${parseColor("bg-decorative1-main")};
    }
  }
  :focus-visible {
    .oak-save-count {
      box-shadow: ${parseDropShadow("drop-shadow-centered-lemon")},
        ${parseDropShadow("drop-shadow-centered-grey")};
    }
  }
`;

export const OakSaveCount = ({ count, href, loading }: OakSaveCountProps) => {
  const iconName = count === 0 ? "bookmark-outlined" : "bookmark-filled";
  const showTruncatedCount = count > 99;
  const containerWidth = showTruncatedCount
    ? "all-spacing-11"
    : "all-spacing-10";

  return (
    <StyledInternalButton as="a" href={href}>
      <OakScreenReader>View my library</OakScreenReader>
      <OakFlex
        $width={["all-spacing-7", containerWidth]}
        $height="all-spacing-7"
        $background={
          loading ? "bg-btn-secondary-hover" : "bg-decorative1-subdued"
        }
        $alignItems="center"
        $justifyContent={["center", "initial"]}
        $pa={["inner-padding-none", "inner-padding-ssx"]}
        $borderRadius="border-radius-s"
        $borderColor={
          loading ? "border-neutral-lighter" : "bg-decorative1-main"
        }
        $ba="border-solid-s"
        className={loading ? undefined : "oak-save-count"}
      >
        <OakIcon
          iconName={iconName}
          data-testid={iconName}
          $width="all-spacing-6"
        />
        <OakBox
          $width="all-spacing-6"
          $textAlign="center"
          $display={["none", "block"]}
        >
          <OakSpan
            $font="heading-light-7"
            aria-label={`${count} saved unit${count === 1 ? "" : "s"}`}
          >
            {showTruncatedCount ? "99+" : count}
          </OakSpan>
        </OakBox>
      </OakFlex>
    </StyledInternalButton>
  );
};
