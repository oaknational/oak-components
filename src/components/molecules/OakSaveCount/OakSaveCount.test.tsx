import React from "react";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { OakSaveCount } from "./OakSaveCount";

const defaultProps = {
  count: 0,
  href: "#",
};

describe("OakSaveCount", () => {
  it("should render a count", () => {
    render(<OakSaveCount {...defaultProps} />);
    const count = screen.getByRole("link");
    expect(count).toBeInTheDocument();
    expect(count).toHaveTextContent("0");
  });
  it("should match snapshot", () => {
    const { container } = render(<OakSaveCount {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
