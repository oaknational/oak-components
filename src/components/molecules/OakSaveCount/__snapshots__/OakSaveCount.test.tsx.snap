// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OakSaveCount should match snapshot 1`] = `
.c3 {
  width: 2rem;
  height: 2rem;
  padding: 0rem;
  background: #dff9de;
  border: 0.063rem solid;
  border-color: #bef2bd;
  border-radius: 0.25rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c5 {
  position: relative;
  width: 1.5rem;
  min-width: 1.5rem;
  height: 2rem;
  min-height: 2rem;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c7 {
  width: 1.5rem;
  display: none;
  font-family: --var(google-font),Lexend,sans-serif;
  text-align: center;
}

.c6 {
  object-fit: contain;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c8 {
  font-family: --var(google-font),Lexend,sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.25rem;
  -webkit-letter-spacing: 0.0115rem;
  -moz-letter-spacing: 0.0115rem;
  -ms-letter-spacing: 0.0115rem;
  letter-spacing: 0.0115rem;
}

.c2 {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  -webkit-clip: rect(0,0,0,0);
  clip: rect(0,0,0,0);
  white-space: nowrap;
  border: 0;
}

.c0 {
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  text-align: left;
  font-family: unset;
  outline: none;
  font-family: --var(google-font),Lexend,sans-serif;
}

.c0:disabled {
  pointer-events: none;
  cursor: default;
}

.c1:hover .oak-save-count {
  background-color: #bef2bd;
}

.c1:focus-visible .oak-save-count {
  box-shadow: 0 0 0 0.125rem rgba(255,229,85,100%), 0 0 0 0.3rem rgba(87,87,87,100%);
}

@media (min-width:750px) {
  .c3 {
    width: 3.5rem;
  }
}

@media (min-width:750px) {
  .c3 {
    padding: 0.25rem;
  }
}

@media (min-width:750px) {
  .c7 {
    display: block;
  }
}

@media (min-width:750px) {
  .c4 {
    -webkit-box-pack: initial;
    -webkit-justify-content: initial;
    -ms-flex-pack: initial;
    justify-content: initial;
  }
}

<div>
  <a
    aria-label="0 saved units"
    class="c0 c1"
    href="#"
  >
    <span
      class="c2"
    >
      View my library
    </span>
    <div
      class="c3 c4 oak-save-count"
    >
      <div
        class="c5"
        data-testid="bookmark-outlined"
      >
        <img
          alt=""
          class="c6"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          src="https://res.cloudinary.com/mock-cloudinary-cloud/image/upload/v1734519491/icons/bookmark-outlined_rxe5v0.svg"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <div
        class="c7"
      >
        <span
          class="c8"
        >
          0
        </span>
      </div>
    </div>
  </a>
</div>
`;
